python/demo/demo_navier_stokes.py

```diff
@@ -0,0 +1,428 @@
+
+
+def domain_average(msh, v):
+    """Compute the average of a function over the domain"""
+    vol = msh.comm.allreduce(
+        fem.assemble_scalar(fem.form(fem.Constant(msh, 1.0) * dx)), op=MPI.SUM)
+    return (1 / vol) * msh.comm.allreduce(fem.assemble_scalar(fem.form(v * dx)), op=MPI.SUM)
+
+
+def u_e_expr(x):
+    """Expression for the exact velocity solution to Kovasznay flow"""
+    return np.vstack((1 - np.exp(
+        (Re / 2 - np.sqrt(Re**2 / 4 + 4 * np.pi**2)) * x[0]) * np.cos(2 * np.pi * x[1]),
+        (Re / 2 - np.sqrt(Re**2 / 4 + 4 * np.pi**2)) / (2 * np.pi) * np.exp(
+            (Re / 2 - np.sqrt(Re**2 / 4 + 4 * np.pi**2)) * x[0])
+        * np.sin(2 * np.pi * x[1])))
+
+
+def p_e_expr(x):
+    """Expression for the exact pressure solution to Kovasznay flow"""
+    return (1 / 2) * (1 - np.exp(2 * (Re / 2 - np.sqrt(Re**2 / 4 + 4 * np.pi**2)) * x[0]))
+
+
+def f_expr(x):
+    """Expression for the applied force"""
+    return np.vstack((np.zeros_like(x[0]),
+                      np.zeros_like(x[0])))
+
+
+def boundary_marker(x):
+    return np.isclose(x[0], 0.0) | np.isclose(x[0], 1.0) | np.isclose(x[1], 0.0) | np.isclose(x[1], 1.0)
+
+
+# We define some simulation parameters
+
+n = 16
+num_time_steps = 25
+t_end = 10
+Re = 25  # Reynolds Number
+k = 1  # Polynomial degree
+
+# Next, we create a mesh and the required functions spaces over it.
+# Since the velocity uses an $H(\textnormal{div})$-conforming function
+# space, we also create a vector valued discontinuous Lagrange space to
+# interpolate into for artifact free visualisation.
+
+msh = mesh.create_unit_square(MPI.COMM_WORLD, n, n)
+
+# Function spaces for the velocity and for the pressure
+V = fem.FunctionSpace(msh, ("Raviart-Thomas", k + 1))
+Q = fem.FunctionSpace(msh, ("Discontinuous Lagrange", k))
+
+# Funcion space for visualising the velocity field
+W = fem.VectorFunctionSpace(msh, ("Discontinuous Lagrange", k + 1))
+
+# Define trial and test functions
+
+u, v = TrialFunction(V), TestFunction(V)
+p, q = TrialFunction(Q), TestFunction(Q)
+
+delta_t = fem.Constant(msh, t_end / num_time_steps)
+alpha = fem.Constant(msh, 6.0 * k**2)
+
+h = CellDiameter(msh)
+n = FacetNormal(msh)
+
+
+def jump(phi, n):
+    return outer(phi("+"), n("+")) + outer(phi("-"), n("-"))
+
+
+# We solve the Stokes problem for the initial condition, so the
+# convective terms are omitted for now
+
+a_00 = (1 / Re) * (inner(grad(u), grad(v)) * dx
+                   - inner(avg(grad(u)), jump(v, n)) * dS
+                   - inner(jump(u, n), avg(grad(v))) * dS
+                   + alpha / avg(h) * inner(jump(u, n), jump(v, n)) * dS
+                   - inner(grad(u), outer(v, n)) * ds
+                   - inner(outer(u, n), grad(v)) * ds
+                   + alpha / h * inner(outer(u, n), outer(v, n)) * ds)
+a_01 = - inner(p, div(v)) * dx
+a_10 = - inner(div(u), q) * dx
+a_11 = fem.Constant(msh, 0.0) * inner(p, q) * dx
```

Comment:
This is rather ugly. I see that it's required to support the Dirichlet condition to constrain the pressure.

Is there a better way? Maybe supporting 'diagonal' matrices from the DOLFINx side, where for block problems the shape can be deduced from the other blocks?

Reply:
In the case of this demo, I'd rather not have to apply  Dirichlet BC to the pressure space at all. MUMPS should be able to deal with the small nullspace, I think. I'll see what I can do to tidy it up.

Still, having the functionality you mention would be useful for other applications